### PR TITLE
fix: generated pkg-config (.pc) files on CMake build

### PR DESCRIPTION
**Pull request by luau-project**
## Description

Right after cminpack v1.3.9 was released, I opened a pull request on MSYS2 to insert cminpack on their repositories [https://github.com/msys2/MINGW-packages/pull/21025](https://github.com/msys2/MINGW-packages/pull/21025). During the review process, the reviewer gave some valuable inputs:

1. Asked if there was a way to avoid the conversion from LF to CRLF on Windows (which I addressed on [https://github.com/devernay/cminpack/pull/63](https://github.com/devernay/cminpack/pull/63));
2. When building the static library through CMake, cminpack adds ```_s``` suffix to the library name. The reviewer noticed that the generated pkg-config files (*.pc) didn't match well the generated static library files. Investigating further, I noticed that all the installed .pc files had a wrong link flag (-l<libname>).

Thus, this pull request addresses the second point above and fixes the installed .pc files.

### The current released version (v1.3.9.tar.gz)

To have a clear understand of what is happening in the actual version of cminpack (v1.3.9), I prepared a simple script to be run on Debian-based distros (Debian 12 stable, Ubuntu 22.04 +, etc).

1. Install the tooling:
```bash
sudo apt install -y gcc make git cmake
```

2. Build, test and install the static version of cminpack v1.3.9 from the released tarball

```bash
cd /tmp
wget https://github.com/devernay/cminpack/archive/v1.3.9.tar.gz
tar -xf v1.3.9.tar.gz
mkdir build-v1.3.9
cd build-v1.3.9/
cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=OFF -DUSE_BLAS=OFF --install-prefix $PWD/static -S ../cminpack-1.3.9/ -B _static
cmake --build _static --config Release
ctest --test-dir _static/ -C Release
cmake --install _static/ --config Release
```
3. If you ```cat static/lib/pkgconfig/cminpack.pc``` the installed pkg-config file for the double-precision version, you get the following screen:
Screenshot from 2024-05-30 19-21-23
4. As you can see in the image above, the -l flag on the pkg-config file for the double-precision version has ```-lcminpack```, but the file is shown as ```/tmp/build-v1.3.9/static/lib/pkgconfig/libcminpack_s.a```, which is an issue;
5. The same kind of issue also happens for the single and long-double precision versions:
Screenshot from 2024-05-30 19-21-44

### The fixed version on my fork

1. Build, test and install the static version of cminpack from my forked repo branch

```bash
cd /tmp
git clone --branch=fix-pkg-config https://github.com/luau-project/cminpack.git cminpack-fork
mkdir fork-v1.3.9
cd fork-v1.3.9/
cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=OFF -DUSE_BLAS=OFF --install-prefix $PWD/static -S ../cminpack-fork/ -B _static
cmake --build _static/ --config Release
ctest --test-dir _static/ -C Release
cmake --install _static/ --config Release
```
2. I decided to rename the pkg-config file to match the installed library name. So, if you ```cat static/lib/pkgconfig/cminpack_s.pc``` the installed pkg-config file for the double-precision version, you get the following screen:
Screenshot from 2024-05-30 19-29-05
3. As you can see in the image above, the -l flag on the pkg-config file for the double-precision version has ```-lcminpack_s```, which is correct compared to the file shown ```/tmp/build-v1.3.9/static/lib/pkgconfig/libcminpack_s.a```;
4. The same fix was applied for all other precisions as you can see:
Screenshot from 2024-05-30 19-29-30

## How

1. I had to make a few changes regarding the suffix. In the old setup, it was used config flags like ```CMAKE_RELEASE_POSTFIX```. I changed the way how it works in the [CMakeLists.txt](CMakeLists.txt) and propagated all the changes down to tests on [examples/CMakeLists.txt](examples/CMakeLists.txt);
2. I also added a guard to only test the library if the double precision version is being built. Currently, the tests are only implemented for the double precision version. These changes were employed on  [examples/CMakeLists.txt](examples/CMakeLists.txt);

## Conclusion

I tested locally the shared and static versions. In my view, everything is working as expected now.

### CODE REVIEW

**Comment by devernay** [comment]
Nice!
Can you please confirm that double-precision shared minpack will be unaffected by this? My fear is that some dependencies may depend on this, so we shouldn't change the package name (cminpack.pc).

File: CMakeLists.txt [hunk 1]
```diff
@@ -142,49 +155,61 @@ foreach (cminpack_lib ${cminpack_libs})
...
   install (FILES ${cminpack_hdrs} DESTINATION ${CMINPACK_INCLUDE_INSTALL_DIR}
     COMPONENT cminpack_hdrs)
 
-  set_target_properties (${cminpack_lib} PROPERTIES
-    POSITION_INDEPENDENT_CODE ON
-    VERSION ${CMINPACK_VERSION}
-    SOVERSION ${CMINPACK_SOVERSION}
+  if (USE_BLAS AND BLAS_FOUND)
+    set_target_properties (${cminpack_lib} PROPERTIES
+      LINK_FLAGS "${BLAS_LINKER_FLAGS}"
```
Comment (jschueller) [inline CMakeLists.txt:175]: you use target_link_options instead (bump cmake version to 3.13 which is old enough)

File: CMakeLists.txt [hunk 2]
```diff
@@ -142,49 +155,61 @@ foreach (cminpack_lib ${cminpack_libs})
...
-    if (${cminpack_lib} STREQUAL cminpackld)
+    if (${cminpack_lib_name} STREQUAL cminpackld)
       message (WARNING "BLAS cannot be used for the extended precision version")
     else ()
       find_package (BLAS)
       if (BLAS_FOUND)
         target_link_libraries (${cminpack_lib} PUBLIC ${BLAS_LIBRARIES})
-        set_target_properties (${cminpack_lib} PROPERTIES LINK_FLAGS "${BLAS_LINKER_FLAGS}")
         target_compile_definitions (${cminpack_lib} PUBLIC USE_BLAS)
+        list(APPEND PC_CMINPACK_CFLAGS "-DUSE_BLAS")
```
Comment (jschueller) [inline CMakeLists.txt]: thats not a required public define, only private

File: CMakeLists.txt [hunk 3]
```diff
@@ -78,41 +78,54 @@ elseif (${CMINPACK_PRECISION} STREQUAL all)
...
+
+set(cminpack_lib_suffix "")
+
 if (BUILD_SHARED_LIBS)
   message (STATUS "Building shared libraries.")
-  if (NOT DEFINED CMAKE_DEBUG_POSTFIX)
-    set (CMAKE_DEBUG_POSTFIX _d)
-  endif ()
+  if ("${CMAKE_BUILD_TYPE}" MATCHES "Debug")
+   set(cminpack_lib_suffix "_d")
```
Comment (jschueller) [inline CMakeLists.txt:87]: whats the rationale behind this and not relying on the global settings ?
sometimes you want to debug but not change the library name in order not to have to relink everything

**Comment by luau-project** [comment]
Hello @devernay .

I saw that there were many commits in the last few days, and I will need some time to understand what happened.

Hello @jschueller .

I saw that you asked a few questions about chunks of scripts on the CMakeLissts.txt file. If you follow the diff, you can see that I didn't change many things, I mostly moved things to different places, trying to preserve the maximum amount of work that was done by the people who implemented this CMake script in the past.

In short, I agree with you that the current CMake script is a bit confuse, but it was just derived from past work.

In my honest opinion, I think the CMakeLists.txt script should be reworked from scratch, because it is hard to maintain and to add features.

I'll work on a complete rewrite of the CMake build in my fork. However, this will not happen for now.

**Comment by devernay** [comment]
my changes were simply fixing BLAS support in the code

**Comment by devernay** [comment]
Should we close this one in favor of #71 ?

**Comment by luau-project** [comment]
> Should we close this one in favor of #71 ?

In my view, yes. #71 seems better.
